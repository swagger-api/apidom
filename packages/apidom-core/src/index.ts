import { has } from 'ramda';
import { isPlainObject, isString } from 'ramda-adjunct';
import { Element, Namespace as INamespace } from 'minim';

import defaultNamespaceInstance from './namespace';
import serializeValue from './serializers/value';
import serializeJSON from './serializers/json';
import serializeYAML from './serializers/yaml-1-2';

export { dispatchPlugins as dispatchRefractorPlugins } from './refractor/plugins/utils';
export { default as refractorPluginElementIdentity } from './refractor/plugins/element-identity';
export { default as refractorPluginSemanticElementIdentity } from './refractor/plugins/semantic-element-identity';

export { default as MediaTypes } from './media-types';

export { Element, MemberElement, KeyValuePair, ObjectSlice, ArraySlice, refract } from 'minim';
export type { NamespacePluginOptions, Attributes, Meta } from 'minim';
export type { PositionRange, Position } from './elements/SourceMap';
export { default as namespace, Namespace, createNamespace } from './namespace';

export {
  ObjectElement,
  ArrayElement,
  BooleanElement,
  NullElement,
  NumberElement,
  StringElement,
  LinkElement,
  RefElement,
  AnnotationElement,
  CommentElement,
  ParseResultElement,
  SourceMapElement,
} from './refractor/registration';

export {
  isElement,
  isStringElement,
  isNumberElement,
  isNullElement,
  isBooleanElement,
  isArrayElement,
  isObjectElement,
  isMemberElement,
  isLinkElement,
  isRefElement,
  isAnnotationElement,
  isParseResultElement,
  isSourceMapElement,
  isPrimitiveElement,
  hasElementSourceMap,
  includesSymbols,
  includesClasses,
} from './predicates/index';
export { default as createPredicate } from './predicates/helpers';

export { filter, reject, find, findAtOffset, some, traverse, parents } from './traversal/index';
export {
  visit,
  BREAK,
  mergeAllVisitors,
  getNodeType,
  keyMapDefault as keyMap,
} from './traversal/visitor';
export { transclude, default as Transcluder } from './transcluder/index';
export { dereference } from './util';
export { cloneShallow, cloneDeep } from './clone';
export { default as CloneError } from './clone/errors/CloneError';
export { default as DeepCloneError } from './clone/errors/DeepCloneError';
export { default as ShallowCloneError } from './clone/errors/ShallowCloneError';

/**
 * Transforms data to an Element from a particular namespace.
 */
export const from = (data: any, namespace: INamespace = defaultNamespaceInstance): Element => {
  if (isString(data)) {
    // JSON serialized refract
    try {
      return namespace.fromRefract(JSON.parse(data));
    } catch {
      // noop
    }
  }
  if (isPlainObject(data) && has('element', data)) {
    // refract javascript structure
    return namespace.fromRefract(data);
  }

  return namespace.toElement(data);
};

/**
 * Transforms the ApiDOM into JavaScript POJO.
 * This POJO would be the result of interpreting the ApiDOM
 * into JavaScript structure.
 */
export const toValue = serializeValue;

/**
 * Transforms the ApiDOM into JSON string.
 */
export const toJSON = serializeJSON;

/**
 * Transforms the ApiDOM into YAML string.
 */
export const toYAML = serializeYAML;

/**
 * Creates a refract representation of an Element.
 * https://github.com/refractproject/refract-spec
 */
export const dehydrate = (
  element: Element,
  namespace: INamespace = defaultNamespaceInstance,
): any => {
  return namespace.toRefract(element);
};

/**
 * Create a refracted string representation of an Element.
 */
export const toString = (
  element: Element,
  namespace: INamespace = defaultNamespaceInstance,
): string => {
  const refractStructure = dehydrate(element, namespace);
  return JSON.stringify(refractStructure);
};

export { default as sexprs } from './sexprs';
export { default as deepmerge } from './deepmerge';
