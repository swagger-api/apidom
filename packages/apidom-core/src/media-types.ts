import { NotImplementedError } from '@swagger-api/apidom-error';

class MediaTypes<T> extends Array<T> {
  unknownMediaType = 'application/octet-stream';

  // eslint-disable-next-line class-methods-use-this
  filterByFormat() {
    throw new NotImplementedError('"filterByFormat" method is not implemented.');
  }

  // eslint-disable-next-line class-methods-use-this
  findBy() {
    throw new NotImplementedError('"findBy" method is not implemented.');
  }

  // eslint-disable-next-line class-methods-use-this
  latest() {
    throw new NotImplementedError('"latest" method is not implemented.');
  }
}

export default MediaTypes;
