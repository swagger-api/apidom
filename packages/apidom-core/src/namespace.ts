import { Namespace as MinimNamespace, NamespacePlugin } from 'minim';
import { isPlainObject } from 'ramda-adjunct';

import AnnotationElement from './elements/Annotation.ts';
import CommentElement from './elements/Comment.ts';
import ParseResultElement from './elements/ParseResult.ts';
import SourceMapElement from './elements/SourceMap.ts';

/**
 * @public
 */
export class Namespace extends MinimNamespace {
  constructor() {
    super();

    this.register('annotation', AnnotationElement);
    this.register('comment', CommentElement);
    this.register('parseResult', ParseResultElement);
    this.register('sourceMap', SourceMapElement);
  }
}

/**
 * @public
 */
const namespace = new Namespace();

/**
 * @public
 */
export const createNamespace = (namespacePlugin?: NamespacePlugin): Namespace => {
  const namespaceInstance = new Namespace();

  if (isPlainObject(namespacePlugin)) {
    namespaceInstance.use(namespacePlugin);
  }

  return namespaceInstance;
};

export default namespace;
