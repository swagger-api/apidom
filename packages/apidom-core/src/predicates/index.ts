import {
  Element,
  StringElement,
  NumberElement,
  NullElement,
  BooleanElement,
  ArrayElement,
  ObjectElement,
  MemberElement,
  LinkElement,
  RefElement,
} from 'minim';
import { all } from 'ramda';
import { included } from 'ramda-adjunct';

import AnnotationElement from '../elements/Annotation';
import CommentElement from '../elements/Comment';
import ParserResultElement from '../elements/ParseResult';
import SourceMapElement from '../elements/SourceMap';
import createPredicate, { isElementType as isElementTypeHelper } from './helpers';

export const isElement = createPredicate(({ hasBasicElementProps, primitiveEq }) => {
  return (element: any) =>
    element instanceof Element ||
    (hasBasicElementProps(element) && primitiveEq(undefined, element));
});

export const isStringElement = createPredicate(({ hasBasicElementProps, primitiveEq }) => {
  return (element: any) =>
    element instanceof StringElement ||
    (hasBasicElementProps(element) && primitiveEq('string', element));
});

export const isNumberElement = createPredicate(({ hasBasicElementProps, primitiveEq }) => {
  return (element: any) =>
    element instanceof NumberElement ||
    (hasBasicElementProps(element) && primitiveEq('number', element));
});

export const isNullElement = createPredicate(({ hasBasicElementProps, primitiveEq }) => {
  return (element: any) =>
    element instanceof NullElement ||
    (hasBasicElementProps(element) && primitiveEq('null', element));
});

export const isBooleanElement = createPredicate(({ hasBasicElementProps, primitiveEq }) => {
  return (element: any) =>
    element instanceof BooleanElement ||
    (hasBasicElementProps(element) && primitiveEq('boolean', element));
});

export const isObjectElement = createPredicate(
  ({ hasBasicElementProps, primitiveEq, hasMethod }) => {
    return (element: any) =>
      element instanceof ObjectElement ||
      (hasBasicElementProps(element) &&
        primitiveEq('object', element) &&
        hasMethod('keys', element) &&
        hasMethod('values', element) &&
        hasMethod('items', element));
  },
);

export const isArrayElement = createPredicate(
  ({ hasBasicElementProps, primitiveEq, hasMethod }) => {
    return (element: any) =>
      (element instanceof ArrayElement && !(element instanceof ObjectElement)) ||
      (hasBasicElementProps(element) &&
        primitiveEq('array', element) &&
        hasMethod('push', element) &&
        hasMethod('unshift', element) &&
        hasMethod('map', element) &&
        hasMethod('reduce', element));
  },
);

export const isMemberElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof MemberElement ||
      (hasBasicElementProps(element) &&
        isElementType('member', element) &&
        primitiveEq(undefined, element));
  },
);

export const isLinkElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof LinkElement ||
      (hasBasicElementProps(element) &&
        isElementType('link', element) &&
        primitiveEq(undefined, element));
  },
);

export const isRefElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof RefElement ||
      (hasBasicElementProps(element) &&
        isElementType('ref', element) &&
        primitiveEq(undefined, element));
  },
);

export const isAnnotationElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof AnnotationElement ||
      (hasBasicElementProps(element) &&
        isElementType('annotation', element) &&
        primitiveEq('array', element));
  },
);

export const isCommentElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof CommentElement ||
      (hasBasicElementProps(element) &&
        isElementType('comment', element) &&
        primitiveEq('string', element));
  },
);

export const isParseResultElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ParserResultElement ||
      (hasBasicElementProps(element) &&
        isElementType('parseResult', element) &&
        primitiveEq('array', element));
  },
);

export const isSourceMapElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof SourceMapElement ||
      (hasBasicElementProps(element) &&
        isElementType('sourceMap', element) &&
        primitiveEq('array', element));
  },
);

export const isPrimitiveElement = (element: any): boolean => {
  return (
    isElementTypeHelper('object', element) ||
    isElementTypeHelper('array', element) ||
    isElementTypeHelper('boolean', element) ||
    isElementTypeHelper('number', element) ||
    isElementTypeHelper('string', element) ||
    isElementTypeHelper('null', element) ||
    isElementTypeHelper('member', element)
  );
};

export const hasElementSourceMap = (element: any): boolean => {
  return isSourceMapElement(element?.meta?.get?.('sourceMap'));
};

export const includesSymbols = <T extends Element>(symbols: string[], element: T): boolean => {
  if (symbols.length === 0) {
    return true;
  }

  const elementSymbols = element.attributes.get('symbols');

  if (!isArrayElement(elementSymbols)) {
    return false;
  }

  return all(included(elementSymbols.toValue()), symbols);
};

export const includesClasses = <T extends Element>(classes: string[], element: T): boolean => {
  if (classes.length === 0) {
    return true;
  }

  return all(included(element.classes.toValue()), classes);
};
