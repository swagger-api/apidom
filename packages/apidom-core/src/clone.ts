import { ArraySlice, ObjectSlice, KeyValuePair, Element } from 'minim';
import { ApiDOMStructuredError } from '@swagger-api/apidom-error';

import { isElement } from './predicates';

type FinalCloneTypes = KeyValuePair | ArraySlice | ObjectSlice;

const cloneShallowKeyValuePair = (keyValuePair: KeyValuePair) => {
  const { key, value } = keyValuePair;
  return new KeyValuePair(key, value);
};

const cloneShallowArraySlice = (arraySlice: ArraySlice) => {
  const items = [...arraySlice];
  return new ArraySlice(items);
};

const cloneShallowObjectSlice = (objectSlice: ObjectSlice) => {
  const items = [...objectSlice];
  return new ObjectSlice(items);
};

/* eslint-disable no-underscore-dangle */
const cloneShallowElement = <T extends Element>(element: T): T => {
  // @ts-ignore
  const copy = new element.constructor();

  copy.element = element.element;

  if (element.meta.length > 0) {
    copy._meta = cloneShallowElement(element.meta);
  }

  if (element.attributes.length > 0) {
    copy._attributes = cloneShallowElement(element.attributes);
  }

  if (isElement(element.content)) {
    const content = element.content as unknown as Element;
    copy.content = cloneShallowElement(content);
  } else if (Array.isArray(element.content)) {
    copy.content = [...element.content];
  } else if ((element.content as unknown) instanceof KeyValuePair) {
    copy.content = cloneShallowKeyValuePair(element.content);
  } else {
    copy.content = element.content;
  }

  return copy;
};
/* eslint-enable */

export const cloneShallow = <T extends Element | FinalCloneTypes>(value: T): T => {
  if (value instanceof KeyValuePair) {
    return cloneShallowKeyValuePair(value) as T;
  }

  if (value instanceof ObjectSlice) {
    return cloneShallowObjectSlice(value) as T;
  }

  if (value instanceof ArraySlice) {
    return cloneShallowArraySlice(value) as T;
  }

  if (isElement(value)) {
    return cloneShallowElement(value) as T;
  }

  throw new ApiDOMStructuredError("Value provided to cloneShallow function couldn't be cloned", {
    value,
  });
};

const invokeClone = <T extends Element | FinalCloneTypes>(value: T): T => {
  if (typeof value?.clone === 'function') {
    return value.clone() as T;
  }
  return value;
};

export const cloneDeep = <T extends Element | FinalCloneTypes>(value: T): T => {
  if (value instanceof ObjectSlice) {
    const items = [...value].map(invokeClone) as T[];
    return new ObjectSlice(items) as T;
  }

  if (value instanceof ArraySlice) {
    const items = [...value].map(invokeClone) as T[];
    return new ArraySlice(items) as T;
  }

  if (typeof value?.clone === 'function') {
    return value.clone() as T;
  }

  throw new ApiDOMStructuredError("Value provided to cloneDeep function couldn't be cloned", {
    value,
  });
};
