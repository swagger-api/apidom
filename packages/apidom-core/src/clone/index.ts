import { ArraySlice, ObjectSlice, KeyValuePair, Element } from 'minim';

import { isElement } from '../predicates';
import DeepCloneError from './errors/DeepCloneError';
import ShallowCloneError from './errors/ShallowCloneError';

type FinalCloneTypes = KeyValuePair | ArraySlice | ObjectSlice;

const invokeClone = <T extends Element | FinalCloneTypes>(value: T): T => {
  if (typeof value?.clone === 'function') {
    return value.clone() as T;
  }
  return value;
};

export const cloneDeep = <T extends Element | FinalCloneTypes>(value: T): T => {
  if (value instanceof ObjectSlice) {
    const items = [...(value as ObjectSlice)].map(invokeClone) as T[];
    return new ObjectSlice(items) as T;
  }

  if (value instanceof ArraySlice) {
    const items = [...(value as ArraySlice)].map(invokeClone) as T[];
    return new ArraySlice(items) as T;
  }

  if (typeof value?.clone === 'function') {
    return value.clone() as T;
  }

  throw new DeepCloneError("Value provided to cloneDeep function couldn't be cloned", {
    value,
  });
};
cloneDeep.safe = <T>(value: T): T => {
  try {
    return cloneDeep(value as any) as T;
  } catch {
    return value;
  }
};

const cloneShallowKeyValuePair = (keyValuePair: KeyValuePair) => {
  const { key, value } = keyValuePair;
  return new KeyValuePair(key, value);
};

const cloneShallowArraySlice = (arraySlice: ArraySlice) => {
  const items = [...arraySlice];
  return new ArraySlice(items);
};

const cloneShallowObjectSlice = (objectSlice: ObjectSlice) => {
  const items = [...objectSlice];
  return new ObjectSlice(items);
};

/* eslint-disable no-underscore-dangle */
const cloneShallowElement = <T extends Element>(element: T): T => {
  // @ts-ignore
  const copy = new element.constructor();

  copy.element = element.element;

  if (element.meta.length > 0) {
    copy._meta = cloneDeep(element.meta);
  }

  if (element.attributes.length > 0) {
    copy._attributes = cloneDeep(element.attributes);
  }

  if (isElement(element.content)) {
    const content = element.content as unknown as Element;
    copy.content = cloneShallowElement(content);
  } else if (Array.isArray(element.content)) {
    copy.content = [...element.content];
  } else if ((element.content as unknown) instanceof KeyValuePair) {
    copy.content = cloneShallowKeyValuePair(element.content);
  } else {
    copy.content = element.content;
  }

  return copy;
};
/* eslint-enable */

export const cloneShallow = <T extends Element | FinalCloneTypes>(value: T): T => {
  if (value instanceof KeyValuePair) {
    return cloneShallowKeyValuePair(value) as T;
  }

  if (value instanceof ObjectSlice) {
    return cloneShallowObjectSlice(value) as T;
  }

  if (value instanceof ArraySlice) {
    return cloneShallowArraySlice(value) as T;
  }

  if (isElement(value)) {
    return cloneShallowElement(value) as T;
  }

  throw new ShallowCloneError("Value provided to cloneShallow function couldn't be cloned", {
    value,
  });
};
cloneShallow.safe = <T>(value: T): T => {
  try {
    return cloneShallow(value as any) as T;
  } catch {
    return value;
  }
};
