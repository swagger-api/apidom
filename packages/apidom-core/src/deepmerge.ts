import { ObjectElement, ArrayElement, MemberElement, Element } from 'minim';

import { isObjectElement, isArrayElement } from './predicates';

/* eslint-disable @typescript-eslint/no-use-before-define */

type ObjectOrArrayElement = ObjectElement | ArrayElement;
type AnyElement = ObjectElement | ArrayElement | Element;

const emptyElement = (element: ObjectElement | ArrayElement) => {
  const meta = element.meta.clone();
  const attributes = element.attributes.clone();

  // @ts-ignore
  return new element.constructor(undefined, meta, attributes);
};

const cloneMemberElement = (memberElement: MemberElement) =>
  new MemberElement(
    memberElement.key,
    memberElement.value,
    memberElement.meta.clone(),
    memberElement.attributes.clone(),
  );

const cloneUnlessOtherwiseSpecified = (element: AnyElement): AnyElement =>
  isObjectElement(element) || isArrayElement(element)
    ? deepmerge(emptyElement(element as ObjectOrArrayElement), element as ObjectOrArrayElement)
    : element;

const mergeArrayElement = (
  targetElement: ArrayElement,
  sourceElement: ArrayElement,
): ArrayElement =>
  targetElement.concat(sourceElement)['fantasy-land/map'](cloneUnlessOtherwiseSpecified);

const mergeObjectElement = (
  targetElement: ObjectElement,
  sourceElement: ObjectElement,
): ObjectElement => {
  const destination = isObjectElement(targetElement)
    ? emptyElement(targetElement)
    : emptyElement(sourceElement);

  if (isObjectElement(targetElement)) {
    targetElement.forEach((value, key, member) => {
      const clonedMember = cloneMemberElement(member as MemberElement);
      clonedMember.value = cloneUnlessOtherwiseSpecified(value);
      destination.content.push(clonedMember);
    });
  }

  sourceElement.forEach((value, key, member) => {
    const keyValue = key.toValue();
    let clonedMember;

    if (
      isObjectElement(targetElement) &&
      targetElement.hasKey(keyValue) &&
      (isObjectElement(value) || isArrayElement(value))
    ) {
      const targetValue = targetElement.get(keyValue);
      clonedMember = cloneMemberElement(member as MemberElement);
      clonedMember.value = deepmerge(targetValue, value as ObjectOrArrayElement);
    } else {
      clonedMember = cloneMemberElement(member as MemberElement);
      clonedMember.value = cloneUnlessOtherwiseSpecified(value as AnyElement);
    }

    destination.remove(keyValue);
    destination.content.push(clonedMember);
  });

  return destination;
};

export default function deepmerge(
  targetElement: ObjectOrArrayElement,
  sourceElement: ObjectOrArrayElement,
) {
  const sourceIsArrayElement = isArrayElement(sourceElement) && !isObjectElement(sourceElement);
  const targetIsArrayElement = isArrayElement(targetElement) && !isObjectElement(targetElement);
  const sourceAndTargetTypesMatch = sourceIsArrayElement === targetIsArrayElement;

  if (!sourceAndTargetTypesMatch) {
    return cloneUnlessOtherwiseSpecified(sourceElement);
  }

  if (sourceIsArrayElement) {
    return mergeArrayElement(targetElement as ArrayElement, sourceElement as ArrayElement);
  }

  return mergeObjectElement(targetElement as ObjectElement, sourceElement as ObjectElement);
}
/* eslint-enable @typescript-eslint/no-use-before-define */
