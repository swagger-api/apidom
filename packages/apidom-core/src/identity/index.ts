import { Element, StringElement } from 'minim';
import ShortUniqueId from 'short-unique-id';

import ElementIdentityError from './errors/ElementIdentityError';
import { isElement, isStringElement } from '../predicates';

export class IdentityManager {
  protected readonly uuid: ShortUniqueId;

  protected readonly identityMap: WeakMap<Element, StringElement>;

  constructor({ length = 6 } = {}) {
    this.uuid = new ShortUniqueId({ length });
    this.identityMap = new WeakMap();
  }

  identify<T extends Element>(this: IdentityManager, element: T): StringElement {
    if (!isElement(element)) {
      throw new ElementIdentityError(
        'Cannot not identify the element. `element` is neither structurally compatible nor a subclass of an Element class.',
        {
          value: element,
        },
      );
    }

    // use already assigned identity
    if (
      element.meta.hasKey('id') &&
      isStringElement(element.meta.get('id')) &&
      !element.meta.get('id').equals('')
    ) {
      return element.id;
    }

    // assign identity in immutable way
    if (this.identityMap.has(element)) {
      return this.identityMap.get(element)!;
    }

    // return element identity
    const id = new StringElement(this.generateId());
    this.identityMap.set(element, id);
    return id;
  }

  forget<T extends Element>(element: T): boolean {
    if (this.identityMap.has(element)) {
      this.identityMap.delete(element);
      return true;
    }
    return false;
  }

  generateId() {
    return this.uuid.randomUUID();
  }
}

export const defaultIdentityManager = new IdentityManager();
