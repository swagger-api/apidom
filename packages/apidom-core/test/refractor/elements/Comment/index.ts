import { assert, expect } from 'chai';

import { CommentElement, StringElement, sexprs } from '../../../../src';

describe('refractor', function () {
  context('elements', function () {
    context('Comment', function () {
      context('given JavaScript primitive', function () {
        specify('should refract to semantic ApiDOM tree', function () {
          const commentElement = CommentElement.refract('comment');

          expect(sexprs(commentElement)).toMatchSnapshot();
        });
      });

      context('given generic ApiDOM element', function () {
        specify('should refract to semantic ApiDOM tree', function () {
          const stringElement = new StringElement('comment');
          const commentElement = CommentElement.refract(stringElement);

          expect(sexprs(commentElement)).toMatchSnapshot();
        });

        specify('should retain attributes', function () {
          const stringElement = new StringElement('comment', undefined, { attr: true });
          const commentElement = CommentElement.refract(stringElement);

          assert.isTrue(commentElement.attributes.get('attr').equals(true));
        });

        specify('should retain meta', function () {
          const stringElement = new StringElement('comment', { meta: true }, undefined);
          const commentElement = CommentElement.refract(stringElement);

          assert.isTrue(commentElement.meta.get('meta').equals(true));
        });
      });
    });
  });
});
