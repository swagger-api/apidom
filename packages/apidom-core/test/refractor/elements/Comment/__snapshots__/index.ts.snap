// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`refractor elements Comment given JavaScript primitive should refract to semantic ApiDOM tree 1`] = `(CommentElement)`;

exports[`refractor elements Comment given generic ApiDOM element should refract to semantic ApiDOM tree 1`] = `(CommentElement)`;
